tests/stats.c

```diff
@@ -0,0 +1,95 @@
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+    IN THE SOFTWARE.
+*/
+
+#include "../src/nn.h"
+#include "../src/reqrep.h"
+
+#include "testutil.h"
+
+#define SOCKET_ADDRESS "tcp://127.0.0.1:5555"
+
+int main (int argc, char **argv)
+{
+    int rc;
+    int rep1;
+    int req1;
+
+    /*  Test req/rep with full socket types. */
+    rep1 = test_socket (AF_SP, NN_REP);
+    test_bind (rep1, SOCKET_ADDRESS);
+    req1 = test_socket (AF_SP, NN_REQ);
+    test_connect (req1, SOCKET_ADDRESS);
+    nn_sleep (200);
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ACCEPTED_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ESTABLISHED_CONNECTIONS) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_CURRENT_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_SENT) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_RECEIVED) == 0);
+
+    nn_assert (nn_get_statistic(req1, NN_STAT_ACCEPTED_CONNECTIONS) == 0);
+    nn_assert (nn_get_statistic(req1, NN_STAT_ESTABLISHED_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_CURRENT_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_SENT) == 0);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_RECEIVED) == 0);
+
+    test_send (req1, "ABC");
+    nn_sleep (100);
+
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_SENT) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_SENT) == 3);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_RECEIVED) == 0);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_RECEIVED) == 0);
+
+    test_recv(rep1, "ABC");
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_SENT) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_SENT) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_RECEIVED) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_RECEIVED) == 3);
+
+    test_send (rep1, "OK");
+    test_recv (req1, "OK");
+
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_SENT) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_SENT) == 3);
+    nn_assert (nn_get_statistic(req1, NN_STAT_MESSAGES_RECEIVED) == 1);
+    nn_assert (nn_get_statistic(req1, NN_STAT_BYTES_RECEIVED) == 2);
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_SENT) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_SENT) == 2);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_MESSAGES_RECEIVED) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_BYTES_RECEIVED) == 3);
+
+    test_close (req1);
+
+    nn_sleep (100);
+
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ACCEPTED_CONNECTIONS) == 1);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_ESTABLISHED_CONNECTIONS) == 0);
+    nn_assert (nn_get_statistic(rep1, NN_STAT_CURRENT_CONNECTIONS) == 0);
+
+    test_close (rep1);
+
+    return 0;
+}
+
```

Comment:
Btw, I apologize for the coding style screw ups.  I actually don't like the nanomsg coding style -- mostly because its different from the BSD/illumos coding style that I use everywhere else.  I occasionally make screw ups (like failing to insert whitespace before leading parenthesis.)  I'd actually love to change the style, and add some verification tools (e.g. illumos' cstyle checking program) to ensure that we are conforming to the style...  something we can discuss later.